### ISSUE TITLE
Update to support 1.9

### ISSUE THREAD

**Issue by mibby** [Siguza/VirtualPack#36]
Preparation ticket, as VirtualPack most likely will break with 1.9 - since it has a history of breaking with every version change. :)

**Comment by Siguza** [comment]
Thanks for the heads up.
I'm fairly confident it will break. But that's what I get for building directly on NMS code.

**Comment by mibby** [comment]
Have you thought about abstracting the code to not use NMS?

**Comment by Siguza** [comment]
Yes I have, long ago even.
But I need some way to send raw server packets, because the Bukkit API does not allow the implementation of custom UI stuff.
And I built everything else around that, and changing it would amount to more or less a rewrite of the plugin, which is also a thing I have planned for a long time, but never got around to.

Maybe in 50 years when I'm retired, I will finally be able to do all that stuff I never had time for. ^^

**Comment by mibby** [comment]
Have you looked into using ProtocolLib has a dependency for custom packets?

**Comment by Siguza** [comment]
Yes, but that would still require the use of `IInventory`.
Now, I already built [a tool](https://github.com/Siguza/NBTLib/blob/master/src/main/java/net/drgnome/nbtlib/ClassProxy.java) that would allow me to create a safe wrapper around it, but again, doing that and migrating all my existing code would practically amount to a rewrite.

**Comment by mibby** [comment]
```
[15:22:33] [Server thread/INFO]: [VirtualPack] Enabling VirtualPack v2.8.2
[15:22:33] [Server thread/ERROR]: Error occurred while enabling VirtualPack v2.8.2 (Is it up to date?)
java.lang.NoClassDefFoundError: net/minecraft/server/v1_8_R3/IChatBaseComponent
	at net.drgnome.virtualpack.util.ConfigProxy.setDefs(ConfigProxy.java:137) ~[?:?]
	at net.drgnome.virtualpack.util.ConfigProxy.<init>(ConfigProxy.java:91) ~[?:?]
	at net.drgnome.virtualpack.util.Config.reload(Config.java:27) ~[?:?]
	at net.drgnome.virtualpack.VPlugin.reloadConfig(VPlugin.java:884) ~[?:?]
	at org.bukkit.plugin.java.JavaPlugin.getConfig(JavaPlugin.java:162) ~[spigot-1.9.jar:git-Spigot-7d15d07-c194444]
	at net.drgnome.virtualpack.util.Config.reload(Config.java:27) ~[?:?]
	at net.drgnome.virtualpack.VPlugin.init(VPlugin.java:87) ~[?:?]
	at net.drgnome.virtualpack.VPlugin.onEnable(VPlugin.java:264) ~[?:?]
	at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:291) ~[spigot-1.9.jar:git-Spigot-7d15d07-c194444]
	at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:340) [spigot-1.9.jar:git-Spigot-7d15d07-c194444]
	at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:405) [spigot-1.9.jar:git-Spigot-7d15d07-c194444]
	at org.bukkit.craftbukkit.v1_9_R1.CraftServer.loadPlugin(CraftServer.java:361) [spigot-1.9.jar:git-Spigot-7d15d07-c194444]
	at org.bukkit.craftbukkit.v1_9_R1.CraftServer.enablePlugins(CraftServer.java:321) [spigot-1.9.jar:git-Spigot-7d15d07-c194444]
	at net.minecraft.server.v1_9_R1.MinecraftServer.t(MinecraftServer.java:411) [spigot-1.9.jar:git-Spigot-7d15d07-c194444]
	at net.minecraft.server.v1_9_R1.MinecraftServer.l(MinecraftServer.java:376) [spigot-1.9.jar:git-Spigot-7d15d07-c194444]
	at net.minecraft.server.v1_9_R1.MinecraftServer.a(MinecraftServer.java:331) [spigot-1.9.jar:git-Spigot-7d15d07-c194444]
	at net.minecraft.server.v1_9_R1.DedicatedServer.init(DedicatedServer.java:269) [spigot-1.9.jar:git-Spigot-7d15d07-c194444]
	at net.minecraft.server.v1_9_R1.MinecraftServer.run(MinecraftServer.java:522) [spigot-1.9.jar:git-Spigot-7d15d07-c194444]
	at java.lang.Thread.run(Unknown Source) [?:1.8.0_74]
Caused by: java.lang.ClassNotFoundException: net.minecraft.server.v1_8_R3.IChatBaseComponent
	at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:91) ~[spigot-1.9.jar:git-Spigot-7d15d07-c194444]
	at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:86) ~[spigot-1.9.jar:git-Spigot-7d15d07-c194444]
	at java.lang.ClassLoader.loadClass(Unknown Source) ~[?:1.8.0_74]
	at java.lang.ClassLoader.loadClass(Unknown Source) ~[?:1.8.0_74]
	... 19 more
```

**Comment by mibby** [comment]
Hopefully support will be added soon. Much needed plugin for me. :)

**Comment by Siguza** [comment]
http://dev.bukkit.org/bukkit-plugins/virtualpack/#c1203

**Comment by mibby** [comment]
Wonderful! Compiling right away - will let you know if I run into any issues. :)

**Comment by mibby** [comment]
@Siguza Virtual anvil produces a whole bunch of issues.

```
[19:11:48] [Server thread/INFO]: Habo5 issued server command: /virtual a
[19:11:48] [Server thread/WARN]: java.lang.NullPointerException
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.ContainerAnvil.getBukkitView(ContainerAnvil.java:373)
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.ContainerAnvil.getBukkitView(ContainerAnvil.java:15)
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.EntityPlayer.a(EntityPlayer.java:849)
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.Container.addSlotListener(Container.java:59)
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.ContainerAnvil.addSlotListener(ContainerAnvil.java:289)
[19:11:48] [Server thread/WARN]: 	at net.drgnome.virtualpack.util.Util.openWindow(Util.java:470)
[19:11:48] [Server thread/WARN]: 	at net.drgnome.virtualpack.VPack.openAnvil(VPack.java:1295)
[19:11:48] [Server thread/WARN]: 	at net.drgnome.virtualpack.VCommands.anvil(VCommands.java:1699)
[19:11:48] [Server thread/WARN]: 	at net.drgnome.virtualpack.VCommands.tools(VCommands.java:183)
[19:11:48] [Server thread/WARN]: 	at net.drgnome.virtualpack.VCommands.cmd(VCommands.java:120)
[19:11:48] [Server thread/WARN]: 	at net.drgnome.virtualpack.VCommands.main(VCommands.java:281)
[19:11:48] [Server thread/WARN]: 	at net.drgnome.virtualpack.VCommands.cmd(VCommands.java:100)
[19:11:48] [Server thread/WARN]: 	at net.drgnome.virtualpack.VCommands.onCommand(VCommands.java:77)
[19:11:48] [Server thread/WARN]: 	at com.sk89q.bukkit.util.DynamicPluginCommand.execute(DynamicPluginCommand.java:54)
[19:11:48] [Server thread/WARN]: 	at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:146)
[19:11:48] [Server thread/WARN]: 	at org.bukkit.craftbukkit.v1_9_R1.CraftServer.dispatchCommand(CraftServer.java:662)
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.PlayerConnection.handleCommand(PlayerConnection.java:1339)
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.PlayerConnection.a(PlayerConnection.java:1151)
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.PacketPlayInChat.a(PacketPlayInChat.java:45)
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.PacketPlayInChat.a(PacketPlayInChat.java:5)
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.PlayerConnectionUtils$1.run(SourceFile:13)
[19:11:48] [Server thread/WARN]: 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[19:11:48] [Server thread/WARN]: 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.SystemUtils.a(SourceFile:45)
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.MinecraftServer.D(MinecraftServer.java:775)
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.DedicatedServer.D(DedicatedServer.java:404)
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.MinecraftServer.C(MinecraftServer.java:714)
[19:11:48] [Server thread/WARN]: 	at net.minecraft.server.v1_9_R1.MinecraftServer.run(MinecraftServer.java:613)
[19:11:48] [Server thread/WARN]: 	at java.lang.Thread.run(Thread.java:745)
[19:11:48] [Server thread/WARN]: [ChestCommands] Task #84 for ChestCommands v3.1.3 generated an exception
java.lang.NullPointerException
	at net.minecraft.server.v1_9_R1.ContainerAnvil.getBukkitView(ContainerAnvil.java:373) ~[patched_null.jar:git-Paper-452]
	at net.minecraft.server.v1_9_R1.ContainerAnvil.getBukkitView(ContainerAnvil.java:15) ~[patched_null.jar:git-Paper-452]
	at org.bukkit.craftbukkit.v1_9_R1.entity.CraftHumanEntity.getOpenInventory(CraftHumanEntity.java:172) ~[patched_null.jar:git-Paper-452]
	at com.gmail.filoghost.chestcommands.task.RefreshMenusTask.run(RefreshMenusTask.java:20) ~[?:?]
	at org.bukkit.craftbukkit.v1_9_R1.scheduler.CraftTask.run(CraftTask.java:58) ~[patched_null.jar:git-Paper-452]
	at org.bukkit.craftbukkit.v1_9_R1.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:352) [patched_null.jar:git-Paper-452]
	at net.minecraft.server.v1_9_R1.MinecraftServer.D(MinecraftServer.java:784) [patched_null.jar:git-Paper-452]
	at net.minecraft.server.v1_9_R1.DedicatedServer.D(DedicatedServer.java:404) [patched_null.jar:git-Paper-452]
	at net.minecraft.server.v1_9_R1.MinecraftServer.C(MinecraftServer.java:714) [patched_null.jar:git-Paper-452]
	at net.minecraft.server.v1_9_R1.MinecraftServer.run(MinecraftServer.java:613) [patched_null.jar:git-Paper-452]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_74]
```

**Comment by mibby** [comment]
Even crashed the server.

```
---- Minecraft Crash Report ----
// My bad.

Time: 3/5/16 7:17 PM
Description: Exception in server tick loop

java.lang.NullPointerException: Exception in server tick loop
    at net.minecraft.server.v1_9_R1.ContainerAnvil.getBukkitView(ContainerAnvil.java:373)
    at net.minecraft.server.v1_9_R1.ContainerAnvil.getBukkitView(ContainerAnvil.java:15)
    at org.bukkit.craftbukkit.v1_9_R1.event.CraftEventFactory.handleInventoryCloseEvent(CraftEventFactory.java:857)
    at net.minecraft.server.v1_9_R1.PlayerList.disconnect(PlayerList.java:397)
    at net.minecraft.server.v1_9_R1.PlayerConnection.a(PlayerConnection.java:1026)
    at net.minecraft.server.v1_9_R1.NetworkManager.handleDisconnection(NetworkManager.java:317)
    at net.minecraft.server.v1_9_R1.ServerConnection.c(ServerConnection.java:175)
    at net.minecraft.server.v1_9_R1.MinecraftServer.D(MinecraftServer.java:876)
    at net.minecraft.server.v1_9_R1.DedicatedServer.D(DedicatedServer.java:404)
    at net.minecraft.server.v1_9_R1.MinecraftServer.C(MinecraftServer.java:714)
    at net.minecraft.server.v1_9_R1.MinecraftServer.run(MinecraftServer.java:613)
    at java.lang.Thread.run(Thread.java:745)
```

**Comment by Siguza** [comment]
As of f93720098fe44703d26ff7e1d83758c04a72d028, the anvil should be fully functional again.

**Comment by mibby** [comment]
@Siguza Anvils unfortunately still seem to be broken. Can duplicate books, tools, and xp.

Enchant anything with the virtual anvil. It doesn't let you take out the resulting item but you get the book back. But the book is also still in the anvil.

**Comment by Siguza** [comment]
Should all be fixed as of 8e8e4195bc485cef68eca4ebdf18d83a06cfa67a.

**Comment by mibby** [comment]
Apologies for the delay. All has seemed fix since that commit. :) I'll be closing this ticket now.
